### ISSUE TITLE
[SEC] nim-libp2p - Infinite Loop on Forbidden EC Point

### ISSUE THREAD

**Issue by eschorn1** [status-im/nim-libp2p#292]
Tags: nbc-audit-2020-0, difficulty:medium, severity:medium, bug [I don't have perms to add labels]

## Description

When the Curve25519 `mulgen()` function (EC scalar `*` generator) calculates a forbidden point, the code will enter an infinite loop resulting in denial of service.

The `mulgen()` function from [`curve25519.nim`](https://github.com/status-im/nim-libp2p/blob/nbc-audit-2020-0/libp2p/crypto/curve25519.nim#L83) is exported, and is also used to calculate the public key corresponding to a private key in the exported `public()` function. This functionality is relevant to the Noise protocol. Lines 100-102 shown below compare the calculated result to a set of forbidden points that [should be rejected](http://cr.yp.to/ecdh.html#validate). However, if any of these points are encountered the code will enter an infinite loop.

~~~nim
proc mulgen*(_: type[Curve25519], dst: var Curve25519Key, point: Curve25519Key) =
  let defaultBrEc = brEcGetDefault()

  var
    rpoint = point
  rpoint.byteswap()

  block iterate:
    while true:
      block derive:
        let
          size = defaultBrEc.mulgen(
            cast[pcuchar](addr dst[0]),
            cast[pcuchar](addr rpoint[0]),
            Curve25519KeySize,
            EC_curve25519)
        assert size == Curve25519KeySize
        for forbid in ForbiddenCurveValues:
          if dst == forbid:
            break derive
        break iterate
~~~

## Exploit Scenario

Calling `mulgen()` with a scalar that results in a forbidden point will cause an infinite loop.

## Mitigation Recommendation

Adapt `mulgen()` to detect and report an error when a forbidden point is encountered. The error would be handled by the caller, which might choose to generate a different private key.

**Comment by sinkingsugar** [comment]
Hmm, I'm actually confused why this is in a loop anyway!
I think maybe I was using it for something else at some point.

**Comment by sinkingsugar** [comment]
Fixed in #293

**Comment by eschorn1** [comment]
I have reviewed https://github.com/status-im/nim-libp2p/pull/293/files and observe that the loop has been removed and the logic around forbidden points fixed. This addresses the finding and thus this issued can be closed. Thank you!
